free/src/main/scala/cats/free/FreeApplicative.scala

```diff
@@ -4,82 +4,196 @@ package free
+      var argF: FA[F, Any] = argsF.head
+      argsF = argsF.tail
+      argsFLength -= 1
+
+      // rip off every `Ap` in `argF`, peeling off left-associated prefixes
+      if (argF.isInstanceOf[Ap[F, _, _]]) {
+        val lengthInitial = argsFLength
+        // reassociate the functions into a single fn,
+        // and move the arguments into argsF
+        do {
+          val Ap(fn, fp) = argF
+          argsF ::= fp.asInstanceOf[FA[F, Any]]
+          argsFLength += 1
+          argF = fn.asInstanceOf[FA[F, Any]]
+        } while (argF.isInstanceOf[Ap[F, _, _]])
+        // consecutive `ap` calls have been queued as operations;
+        // argF is no longer an `Ap` node, so the entire topmost left-associated
+        // function application branch has been looped through and we've
+        // moved (`argsFLength` - `lengthInitial`) arguments to the stack, through
+        // (`argsFLength` - `lengthInitial`) `Ap` nodes, so the function on the right
+        // which consumes them all must have (`argsFLength` - `lengthInitial`) arguments
+        val argc = argsFLength - lengthInitial
+        fns ::= Fn[G, Any, Any](foldArg(argF.asInstanceOf[FA[F, Any => Any]], f), argc)
+        fnsLength += 1
+        loop()
+      } else {
+        val argT: G[Any] = foldArg(argF, f)
+        if (fns ne Nil) {
+          // single right-associated function application
+          var fn = fns.head
+          fns = fns.tail
+          fnsLength -= 1
+          var res = G.ap(fn.gab)(argT)
+          if (fn.argc > 1) {
+            // this function has more than 1 argument,
+            // bail out of nested right-associated function application
+            fns ::= Fn(res.asInstanceOf[G[Any => Any]], fn.argc - 1)
+            fnsLength += 1
+            loop()
+          } else {
+            if (fnsLength > 0) {
+              // we've got a nested right-associated `Ap` tree,
+              // so apply as many functions as possible
+              @tailrec
+              def innerLoop(): Unit = {
+                fn = fns.head
+                fns = fns.tail
+                fnsLength -= 1
+                res = G.ap(fn.gab)(res)
+                if (fn.argc > 1) {
+                  fns ::= Fn(res.asInstanceOf[G[Any => Any]], fn.argc - 1)
+                  fnsLength += 1
+                  res = G.ap(fn.gab)(res)
+                }
+                if (fn.argc == 1 && fnsLength > 0) innerLoop()
+              }
+
+              innerLoop()
+            }
+            if (fnsLength == 0) res
+            else loop()
+          }
+        } else argT
+      }
     }
 
+    loop().asInstanceOf[G[A]]
+  }
+  // scalastyle:on method.length
+
+
   /** Interpret/run the operations using the semantics of `Applicative[F]`.
-   * Tail recursive only if `F` provides tail recursive interpretation.
-   */
-  final def fold(implicit F: Applicative[F]): F[A] =
+    * Tail recursive only if `F` provides tail recursive interpretation.
+    */
```

Comment:
another nitpick here, the scala doc indentation is changed away from our standard.